Update 0062-SPAM-spam_protection.md (#1474)
vegaprotocol/specs: protocol/0062-SPAM-spam_protection.md

For SW, we only have governance, so the following two policies will do:

Vote transactions can be rejected if a if a party has less than `spam.protection.voting.min.tokens`. Any governance proposal transaction can be rejected if a party has less than `spam.protection.proposal.min.tokens`. Setting these reasonably high provides some level of protection. 
Any qualified voter can vote `spam.protection.max.votes` times per epoch per active proposal (e.g., if it's `3` then  one initial vote and twice change their mind).

If 3 blocks in a row for filled with spam i.e., parties sending substantially more than 3 votes, let's say 50 votes), 
then the number of required tokens is doubled, up to a maximum of 1600 (if someone pays 1.5 million to spam us for 60 

> seconds so be it).

Reviewer: @ze97286 is this something that's been implemented or should we just remove this from the spec?

Reviewer 2: it is implemented.

For SW, we only have governance, so the following two policies will do:

> Vote proposals can only be done using a lot of tokens (say, 100.0000)

Reviewer: Should we reference the relevant network parameters by name? We have `governance.proposal.*.minProposerBalance` which is expected to be set to non-trivial amounts. 

Or do you mean the vote transactions themselves - in which case it's `governance.proposal.*.minVoterBalance`.

Reviewer: In fact I am wrong, it's `spam.protection.proposal.min.tokens` and `spam.protection.voting.min.tokens` here. Let's do the edit now.

> Vote proposals can only be done using a lot of tokens (say, 100.0000)

Reviewer: ```suggestion
Vote transactions can be rejected if a if a party has less than `spam.protection.voting.min.tokens`. Any governance proposal transaction can be rejected if a party has less than `spam.protection.proposal.min.tokens`. Setting these reasonably high provides some level of protection. 
```

> Any qualified voter can vote three times per epoch per active proposal (i.e., one initial vote and twice change their mind).

Reviewer: ```suggestion
Any qualified voter can vote `spam.protection.max.votes` times per epoch per active proposal (e.g., if it's `3` then  one initial vote and twice change their mind).
```

For SW, we only have governance, so the following two policies will do:

Vote transactions can be rejected if a if a party has less than `spam.protection.voting.min.tokens`. Any governance proposal transaction can be rejected if a party has less than `spam.protection.proposal.min.tokens`. Setting these reasonably high provides some level of protection. 
Any qualified voter can vote `spam.protection.max.votes` times per epoch per active proposal (e.g., if it's `3` then  one initial vote and twice change their mind).

> Initially, a qualified voter required at least 100 tokens (i.e., a value of $1500 taking the last coinlist sale).

Reviewer: ```suggestion
```

> Thus, spamming the network for 1 minute would cost $135.000. 

Reviewer: ```suggestion
```

 pre_block rejected, if there are `<num_votes>` or more votes on the same proposal in the blockchain in this epoch, and
 post_block rejected, if there are `<num_votes>` or more on the same proposal in the blockchain plus earlier in the current block.

> - Any tokenholder that had more than 50% if its post-rejected is banned for max (30 seconds, 1/48 of an epoch) or until the next epoch starts, and all its votes (and only those, no trading related transactions) are immediately rejected. E.g. if the epoch duration is 1 day, then the ban period is 30 minutes. If however the epoch is 10 seconds, then the ban period is 30 seconds (or until the start of the next epoch). The test for 50% of the votes is repeated once the next vote is post-rejected, so a it is possible for a violating party to get banned quite quickly again.

Reviewer 2: A couple of comments:
1. this is not only for voting, it's applicable for proposals, delegation, etc. 
2. what happens once the party gets unbanned? lets say they have voted already 3 times time epoch and the above applied to them so they got banned, half an hour passes and ban is lifted. Can they vote again for a given proposal to which they've already voted 3 times or should they be rejected?

Author: 1. True, governance-related transactions would be more precise.
2. The vote count for that epoch stays, so the fourth vote will now be rejected; it will not cause another ban, as the state of the validators should have caught up now, so spam-protection should reject that vote before it enters a block. If the account manages to sneak some transaction through that needs to be post-rejected (e.g., through a vote on another issue), this post rejection can immediately trigger a new ban through the 50% rule.

Reviewer 2: excellent, thanks.

> - Any tokenholder that had more than 50% if its post-rejected is banned for max (30 seconds, 1/48 of an epoch) or until the next epocvh starts, and all its votes (and only those, no trading related transactions) are immediately rejected. E.g. if the epoch duration is 1 day, then the ban period is 30 minutes. If however the epoch is 10 seconds, then the ban period is 30 seconds (or until the start of the next epoch). The test for 50% of the votes is repeated once the next vote is post-rejected, so a it is possible for a violating party to get banned quite quickly again.

Reviewer: ```suggestion
- Any tokenholder that had more than 50% if its post-rejected is banned for max (30 seconds, 1/48 of an epoch) or until the next epoch starts, and all its votes (and only those, no trading related transactions) are immediately rejected. E.g. if the epoch duration is 1 day, then the ban period is 30 minutes. If however the epoch is 10 seconds, then the ban period is 30 seconds (or until the start of the next epoch). The test for 50% of the votes is repeated once the next vote is post-rejected, so a it is possible for a violating party to get banned quite quickly again.
```

 - It is possible for spam transactions to fill a block (<a name="0062-SPAM-010" href="#0062-SPAM-010">0062-SPAM-010</a>)
 - Parties that continue spamming are blocked and eventually unblocked again  (<a name="0062-SPAM-011" href="#0062-SPAM-011">0062-SPAM-011</a>)
 - Any rejection due to spam protection is reported to the user upon transaction submission detailing which criteria the key exceeded / not met  (<a name="0062-SPAM-013" href="#0062-SPAM-013">0062-SPAM-013</a>)  

>  - If a party is banned for too many voting-rejections, it still can send trading related transactions which are not banned. (<a name="0062-SPAM-014" href="#0062-SPAM-014">0062-SPAM-013</a>)  

Reviewer: Needs new AC code, 13 is already in use.
